File: packages/js/http/src/presets.ts
import type { OpenReceiveAuthorize, OpenReceiveAuthorizeContext } from "./authorize.ts";

// Ready-made `authorize` policies for the two common host shapes. Both build on `ctx.tokenValid` —
// the per-order token validity the handler precomputes — so neither needs the token manager, a
// session store, or any manager wiring. Pass one as `authorize` to `createOpenReceiveHttpHandler`.

/** Options for {@link guestCheckout}. */
export interface GuestCheckoutOptions {
  /**
   * Gate for the Tier-3 `invoice.sweep` admin action (denied by default). Return true to allow —
   * e.g. after checking a shared admin secret on the request. Omit to keep sweep closed.
   */
  readonly allowSweep?: (ctx: OpenReceiveAuthorizeContext) => boolean | Promise<boolean>;
}

/**
 * Policy for an guest-checkout site (no accounts) / paywall: anyone may create a checkout, per-order reads and
 * swap actions are gated on the order's capability token (the httpOnly cookie set on create is
 * enough), and admin sweep is denied unless `allowSweep` opts in.
 */
export function guestCheckout(options?: GuestCheckoutOptions): OpenReceiveAuthorize {
  return async (ctx: OpenReceiveAuthorizeContext): Promise<boolean> => {
    if (ctx.action === "checkout.create") {
      return true;
    }
    if (ctx.action === "invoice.sweep") {
      return (await options?.allowSweep?.(ctx)) ?? false;
    }
    return ctx.tokenValid;
  };
}

/** Options for {@link withUser}. */
export interface WithUserOptions<U> {
  /**
   * Whether `user` owns the order named in `ctx.resource.order_id` (a DB lookup, typically). When
   * omitted, ownership falls back to `ctx.tokenValid`, so a logged-in user who also holds the order
   * token is allowed.
   */
  readonly ownsOrder?: (user: U, ctx: OpenReceiveAuthorizeContext) => boolean | Promise<boolean>;
  /** Whether `user` may run the Tier-3 `invoice.sweep` admin action. Denied when omitted. */
  readonly isAdmin?: (user: U) => boolean;
}

/**
 * Policy for a site with logged-in users: `getUser` resolves the request's user (e.g. from a
 * session). A missing user is denied everything; a present user may always create a checkout, may
 * sweep only when `isAdmin` allows, and may read/act on an order per `ownsOrder` (falling back to
 * the order token via `ctx.tokenValid` when `ownsOrder` is not supplied).
 */
export function withUser<U>(
  getUser: (request: Request) => U | undefined | Promise<U | undefined>,
  options?: WithUserOptions<U>,
): OpenReceiveAuthorize {
  return async (ctx: OpenReceiveAuthorizeContext): Promise<boolean> => {
    const user = await getUser(ctx.request);
    if (!user) {
      return false;
    }
    if (ctx.action === "checkout.create") {
      return true;
    }
    if (ctx.action === "invoice.sweep") {
      return options?.isAdmin?.(user) ?? false;
    }
    return options?.ownsOrder ? await options.ownsOrder(user, ctx) : ctx.tokenValid;
  };
}

File: packages/ruby/openreceive-server/lib/openreceive/server/presets.rb
# frozen_string_literal: true

module OpenReceive
  module Server
    # Ready-made `authorize` policies for the two common host shapes — the Ruby port of the shipped
    # @openreceive/http presets (presets.ts). Both build on the handler-precomputed
    # `context[:token_valid]` (per-order token validity), so neither touches the token manager, a
    # session store, or any manager wiring. Pass one as `authorize:` to RackApp / RequestHandler, or
    # assign it to `config.authorize` in the Rails engine.
    #
    # The callables accept the SAME context hash the handler passes every authorize/rate_limit hook:
    #   { action:, request:, resource:, token:, token_valid:, order_id? }
    module Presets
      module_function

      # Policy for an guest-checkout site (no accounts) / paywall: anyone may prepare and create a
      # checkout (and read order summaries), per-order reads and swap actions are gated on the order's
      # capability token (the httpOnly cookie set on create is enough), and admin sweep is denied
      # unless `allow_sweep` opts in. Byte-identical to the JS `guestCheckout`:
      # checkout.prepare / checkout.create / order.summary → true; invoice.sweep → allow_sweep&.call(ctx)
      # or false; else → ctx[:token_valid].
      #
      #   authorize: OpenReceive::Server::Presets.guest_checkout
      #   authorize: OpenReceive::Server::Presets.guest_checkout(allow_sweep: ->(ctx) { admin?(ctx) })
      def guest_checkout(allow_sweep: nil)
        lambda do |ctx|
          case ctx[:action]
          when "checkout.prepare", "checkout.create", "order.summary"
            true
          when "invoice.sweep"
            allow_sweep ? allow_sweep.call(ctx) : false
          else
            ctx[:token_valid]
          end
        end
      end

      # Policy for a site with logged-in users: `get_user` resolves the request's user (e.g. from a
      # session) — it receives `ctx[:request]`. A missing user is denied everything; a present user may
      # always prepare/create a checkout and read summaries, may sweep only when `is_admin` allows, and
      # may read/act on an order per `owns_order` (falling back to the order token via ctx[:token_valid]
      # when `owns_order` is not supplied). Byte-identical to the JS `withUser`.
      #
      #   authorize: OpenReceive::Server::Presets.with_user(
      #     ->(request) { current_user_for(request) },
      #     owns_order: ->(user, ctx) { user.owns_order?(ctx[:order_id]) },
      #     is_admin:   ->(user) { user.admin? }
      #   )
      def with_user(get_user, owns_order: nil, is_admin: nil)
        lambda do |ctx|
          user = get_user.call(ctx[:request])
          next false unless user

          case ctx[:action]
          when "checkout.prepare", "checkout.create", "order.summary"
            true
          when "invoice.sweep"
            is_admin ? is_admin.call(user) : false
          else
            owns_order ? owns_order.call(user, ctx) : ctx[:token_valid]
          end
        end
      end
    end
  end
end
